File: frontend/src/types/agent.ts
export interface AgentProfileOut {
  code: string
  name: string
  focus: string
  issue_types: string[]
  instruction: string
  enabled: boolean
}

export interface AgentUsageOut {
  code: string
  name: string
  call_count: number
  success_count: number
  failed_count: number
  last_called_at?: string | null
}

export interface ReviewTypeMappingOut {
  review_type: string
  label: string
  agent_codes: string[]
}

export interface AgentOverviewOut {
  agents: AgentProfileOut[]
  type_mappings: ReviewTypeMappingOut[]
  usage: AgentUsageOut[]
}

// === v2.0 新增 ===

export type AgentStatus =
  | 'idle' | 'thinking' | 'working' | 'blocked' | 'error' | 'offline'

export interface AgentRuntimeOut {
  code: string
  name: string
  description: string
  icon: string
  color: string
  category: string
  skills: string[]
  status: AgentStatus
  model: string
  call_count: number
  success_count: number
  failed_count: number
  last_called_at?: string | null
}

export interface AgentRuntimeSummaryOut {
  total: number
  by_category: { category: string; count: number }[]
}

export interface AgentSituationOut {
  online: number
  working: number
  idle: number
  today_calls: number
  spectrum: { bucket: string; count: number }[]
  hotspots: { code: string; name: string; count: number }[]
}

// === v3.0 AgentSkill 升级新增 ===

/**
 * Skill 类型枚举,与后端 BaseSkill.SkillType 对齐
 * - self_improvement: 自我进化类 Skill(沉淀经验/蒸馏规则/评估闸门)
 * - proactive: 主动监测类 Skill(主动巡检/发现潜在问题)
 */
export type SkillType = 'self_improvement' | 'proactive'

/**
 * Skill 元数据,对齐后端 SkillMetaOut schema
 * 由 GET /api/agents/{agent_name}/skills 返回
 */
export interface SkillMetaOut {
  /** Skill 唯一名称,如 code_reviewer.self_improve */
  name: string
  /** Skill 描述(中文,供前端展示) */
  description: string
  /** Skill 类型(self_improvement / proactive) */
  type: SkillType
  /** 是否可手动调用(invocable=true 的 Skill 才会出现在手动调用列表) */
  invocable: boolean
  /** 所属 Agent name(如 code_reviewer) */
  agent_name: string
}

/**
 * Skill 调用入参,对齐后端 SkillInvokeIn schema
 * 用于 POST /api/agents/{agent_name}/skills/{skill_name}/invoke
 */
export interface SkillInvokeIn {
  /** 可选 action(如 evolve / check_proactive),留空则由 Skill 默认 action */
  action?: string
  /** Skill 参数 dict(各 Skill 自定义) */
  params?: Record<string, unknown>
}

/**
 * Skill 调用出参,对齐后端 SkillInvokeOut schema
 */
export interface SkillInvokeOut {
  /** 是否成功 */
  success: boolean
  /** Skill 返回数据(各 Skill 自定义结构) */
  data?: Record<string, unknown> | unknown
  /** 错误信息(success=false 时填充) */
  error?: string
  /** 效果标记(success / no_change / failed,与 agent_skill_record.effect 对齐) */
  effect?: string
  /** 执行耗时(毫秒) */
  duration_ms?: number
  /** agent_skill_record 主键(可用于查询调用记录详情) */
  record_id?: number
}

/**
 * Skill 调用记录,对齐后端 AgentSkillRecordOut schema
 * 由 GET /api/agents/skill-records 返回
 */
export interface AgentSkillRecordOut {
  /** 记录主键 */
  id: number
  /** Agent name(如 code_reviewer) */
  agent_name: string
  /** Skill name(如 code_reviewer.self_improve) */
  skill_name: string
  /** 触发类型(manual / scheduled / event / orchestrator) */
  trigger_type: string
  /** 触发来源描述(如 "api:POST /agents/.../invoke") */
  trigger_source?: string | null
  /** 效果标记(success / no_change / failed) */
  effect?: string | null
  /** 是否成功(由 effect === 'success' 派生) */
  success: boolean
  /** 执行耗时(毫秒) */
  duration_ms?: number | null
  /** 输出摘要(Skill 返回数据的精简描述) */
  output_summary?: string | null
  /** 创建时间(ISO8601 字符串) */
  create_time?: string | null
}

File: backend/app/schemas/agent.py
"""
Agent 中心相关 Pydantic Schema
"""
from typing import Any, Optional

from pydantic import BaseModel, Field, field_validator


class AgentProfileOut(BaseModel):
    """代理画像输出: 镜像 app.ai.multi_agent 中的 ReviewAgentProfile (v1.0 兼容)"""
    code: str
    name: str
    focus: str
    issue_types: list[str]
    instruction: str
    enabled: bool = True


class AgentUsageOut(BaseModel):
    """单代理调用统计"""
    code: str
    name: str
    call_count: int = 0
    success_count: int = 0
    failed_count: int = 0
    last_called_at: Optional[str] = None


class ReviewTypeMappingOut(BaseModel):
    """审查类型 → 代理组合映射"""
    review_type: str
    label: str
    agent_codes: list[str]


class AgentOverviewOut(BaseModel):
    """Agent 中心首屏数据聚合 (v1.0 兼容)"""
    agents: list[AgentProfileOut]
    type_mappings: list[ReviewTypeMappingOut]
    usage: list[AgentUsageOut]


# === v2.0 新增 ===


class AgentRuntimeOut(BaseModel):
    """v2.0 真实注册的 Agent 运行时元数据 (与 AgentRegistry 严格同步)"""
    code: str
    name: str
    description: str = ""
    icon: str = "base"
    color: str = "#5B58E8"
    category: str = "general"
    skills: list[str] = []
    status: str = "idle"
    model: str = ""
    call_count: int = 0
    success_count: int = 0
    failed_count: int = 0
    last_called_at: Optional[str] = None

    @field_validator("skills", mode="before")
    @classmethod
    def _normalize_skills(cls, v: Any) -> list[str]:
        """v3.0 起 AgentRegistry 的 skills 为结构化 list[dict]。

        前端契约仍是 string[],故此处兼容性地把结构化技能拍平为技能名,
        避免响应序列化触发 ResponseValidationError。结构化元数据由
        GET /api/agents/{name}/skills 单独提供。
        """
        if not v:
            return []
        return [
            (item.get("name", "") if isinstance(item, dict) else item)
            for item in v
        ]


class AgentRuntimeSummaryOut(BaseModel):
    """注册中心汇总: 总数 + category 分桶"""
    total: int
    by_category: list[dict]


class AgentSituationOut(BaseModel):
    """v2.0 态势感知面板数据"""
    online: int = Field(..., description="当前在岗 Agent 数")
    working: int = Field(0, description="工作中数量(M2 阶段接入)")
    idle: int = Field(0, description="空闲数量")
    today_calls: int = Field(0, description="今日累计调用数")
    spectrum: list[dict] = Field(default_factory=list, description="近 N 分钟调用波形")
    hotspots: list[dict] = Field(default_factory=list, description="近期热点 Agent")


# === v3.0 AgentSkill 升级: Skill 调用相关 Schema ===


class SkillInvokeIn(BaseModel):
    """手动调用 Skill 请求体

    用于 POST /api/agents/{name}/skills/{skill}/invoke 端点,
    admin 通过此接口手动触发任意 Agent 的任意 Skill。

    Attributes:
        action: Skill 子动作(可选), 如 evolve / check_proactive /
                scan_domain / reflect_from_logs
        params: 透传给 Skill 的额外参数(可选)
    """
    action: Optional[str] = Field(
        default=None,
        description="Skill 子动作, 如 evolve/check_proactive/scan_domain/reflect_from_logs",
    )
    params: dict[str, Any] = Field(
        default_factory=dict,
        description="透传给 Skill 的额外参数(action 会合并进去)",
    )


class SkillInvokeOut(BaseModel):
    """手动调用 Skill 响应体

    Attributes:
        success: 是否成功
        data: Skill 调用返回的数据(任意类型)
        error: 失败原因(success=False 时填充)
        effect: 效果标签(success/failed/no_op/proposal_created)
        duration_ms: 执行耗时(毫秒)
        record_id: agent_skill_record 记录 ID
    """
    success: bool
    data: Optional[Any] = None
    error: Optional[str] = None
    effect: str = "success"
    duration_ms: int = 0
    record_id: Optional[int] = None


class AgentSkillRecordOut(BaseModel):
    """Skill 调用记录输出

    供 SkillManager 页面展示调用历史使用。

    Attributes:
        id: 记录 ID
        agent_name: Agent 名称
        skill_name: Skill 名称
        trigger_type: 触发类型(manual/scheduled/event/proactive)
        trigger_source: 触发来源描述
        effect: 效果标签
        success: 是否成功(由 effect == "success" 派生)
        duration_ms: 执行耗时(毫秒)
        output_summary: 输出摘要
        create_time: 创建时间
    """
    id: int
    agent_name: str
    skill_name: str
    trigger_type: str
    trigger_source: str = ""
    effect: str = "success"
    success: bool = True
    duration_ms: int = 0
    output_summary: str = ""
    create_time: Optional[str] = None


class SkillMetaOut(BaseModel):
    """Skill 元数据输出

    供 /api/agents/{name}/skills 列出指定 Agent 的 Skill 元数据使用。

    Attributes:
        name: Skill name(形如 "code_reviewer.self_improve")
        description: Skill 描述
        type: Skill 类型(self_improvement / proactive)
        invocable: 是否可被外部调用
        agent_name: 所属 Agent name
    """
    name: str
    description: str = ""
    type: str = ""
    invocable: bool = True
    agent_name: str = ""
